test/fs_test.go

```diff
@@ -0,0 +1,397 @@
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+)
+
+func TestStateWrapper_OpenInvalidPath(t *testing.T) {
+	st := llb.Scratch().File(llb.Mkfile("/bar", 0644, []byte("hello world")))
+	testEnv.RunTest(context.Background(), t, func(ctx context.Context, gwc gwclient.Client) (*gwclient.Result, error) {
+		rfs := dalec.NewStateRefFS(st, context.Background(), gwc)
+
+		// cannot prefix path with "/", per go path conventions
+		_, err := rfs.Open("/bar")
+		if err == nil {
+			t.Fatal("expected error")
+		}
+
+		assert.True(t, errors.Is(err, fs.ErrInvalid))
+		return rfs.Res()
+	})
+}
+
+func TestStateWrapper_Open(t *testing.T) {
+	st := llb.Scratch().
+		File(llb.Mkfile("/foo", 0644, []byte("hello world")))
+
+	testEnv.RunTest(context.Background(), t, func(ctx context.Context, gwc gwclient.Client) (*gwclient.Result, error) {
+		fs := dalec.NewStateRefFS(st, context.Background(), gwc)
+		f, err := fs.Open("foo")
+		assert.Nil(t, err)
+
+		b := make([]byte, 11)
+		n, err := f.Read(b)
+		assert.Equal(t, err, io.EOF)
+		assert.Equal(t, n, 11)
+
+		return fs.Res()
+	})
+}
+
+func TestStateWrapper_Stat(t *testing.T) {
+	st := llb.Scratch().File(llb.Mkfile("/foo", 0755, []byte("contents")))
+	testEnv.RunTest(context.Background(), t, func(ctx context.Context, gwc gwclient.Client) (*gwclient.Result, error) {
+		rfs := dalec.NewStateRefFS(st, context.Background(), gwc)
+		f, err := rfs.Open("foo")
+		assert.Nil(t, err)
+
+		info, err := f.Stat()
+		assert.Nil(t, err)
+
+		assert.Equal(t, info.IsDir(), false)
+		assert.Equal(t, info.Mode(), fs.FileMode(0755))
+		assert.Equal(t, info.Size(), int64(len([]byte("contents"))))
+		assert.Equal(t, info.Name(), "foo")
+
+		return rfs.Res()
+	})
+}
+
+func TestStateWrapper_ReadDir(t *testing.T) {
+	st := llb.Scratch().File(llb.Mkdir("/bar", 0644)).
+		File(llb.Mkfile("/bar/foo", 0644, []byte("file contents"))).
+		File(llb.Mkdir("/bar/baz", 0644))
+
+	var expectInfo = map[string]struct {
+		perms    fs.FileMode
+		isDir    bool
+		contents []byte
+	}{
+		"/bar/foo": {
+			perms:    0644,
+			isDir:    false,
+			contents: []byte("file contents"),
+		},
+
+		"/bar/baz": {
+			perms: fs.ModeDir | 0644,
+			isDir: true,
+		},
+	}
+
+	testEnv.RunTest(context.Background(), t, func(ctx context.Context, c gwclient.Client) (*gwclient.Result, error) {
+		rfs := dalec.NewStateRefFS(st, ctx, c)
+		root := "/bar"
+		entries, err := rfs.ReadDir(root)
+		assert.Nil(t, err)
+
+		for _, e := range entries {
+			p := path.Join(root, e.Name())
+			want, ok := expectInfo[p]
+			assert.True(t, ok)
+
+			info, err := e.Info()
+			assert.Nil(t, err)
+
+			assert.Equal(t, want.perms, info.Mode())
+			assert.Equal(t, want.perms.String(), info.Mode().String())
+			assert.Equal(t, want.isDir, info.IsDir())
+		}
+
+		return rfs.Res()
+	})
+
+}
+
+func TestStateWrapper_ReadDir_HTTPSource(t *testing.T) {
+	url := "https://patch-diff.githubusercontent.com/raw/kubernetes/kubernetes/pull/120134.patch"
```

Comment:
I'd prefer to not rely on links/git urls in these unit tests if we can avoid it. Are the http and git tests testing different behavior in the FS itself? Because if not, I'm not sure if we need these, since we already have some integration tests where this kind of code would be exercised